Mathlib/Data/Nat/Factorization/Basic.lean

```diff
@@ -651,6 +652,86 @@ theorem factorization_lcm {a b : ℕ} (ha : a ≠ 0) (hb : b ≠ 0) :
+    0 < factorization_lcm_left a b := by
+  apply Nat.pos_of_ne_zero
+  rw [factorization_lcm_left, Finsupp.prod_ne_zero_iff]
+  intro p _ H
+  by_cases h : b.factorization p ≤ a.factorization p
+  · simp only [h, reduceIte, pow_eq_zero_iff', ne_eq] at H
+    simpa [H.1] using H.2
+  · simp only [h, reduceIte, one_ne_zero] at H
+
+lemma factorization_lcm_right_pos :
+    0 < factorization_lcm_right a b := by
+  apply Nat.pos_of_ne_zero
+  rw [factorization_lcm_right, Finsupp.prod_ne_zero_iff]
+  intro p _ H
+  by_cases h : b.factorization p ≤ a.factorization p
+  · simp only [h, reduceIte, pow_eq_zero_iff', ne_eq] at H
+  · simp only [h, ↓reduceIte, pow_eq_zero_iff', ne_eq] at H
+    simpa [H.1] using H.2
+
+lemma coprime_factorization_lcm_left_factorization_lcm_right :
+    (factorization_lcm_left a b).Coprime (factorization_lcm_right a b) := by
+  rw [factorization_lcm_left, factorization_lcm_right]
+  refine coprime_prod_left_iff.mpr fun p hp ↦ coprime_prod_right_iff.mpr fun q hq ↦ ?_
+  dsimp only; split_ifs with h h'
+  any_goals apply coprime_one_left
+  · apply coprime_one_right
+  refine coprime_pow_primes _ _ (prime_of_mem_primeFactors hp) (prime_of_mem_primeFactors hq) ?_
+  contrapose! h'; rwa [← h']
+
+variable {a b}
+
+lemma factorization_lcm_left_mul_factorization_lcm_right (ha : a ≠ 0) (hb : b ≠ 0) :
+    (factorization_lcm_left a b) * (factorization_lcm_right a b) = lcm a b := by
+  rw [← factorization_prod_pow_eq_self (lcm_ne_zero ha hb), factorization_lcm_left,
+    factorization_lcm_right, ← prod_mul]
+  congr; ext p n; split_ifs <;> simp
+
+variable (a b)
+
+lemma factorization_lcm_left_dvd : factorization_lcm_left a b ∣ a := by
+  rcases eq_or_ne a 0 with rfl | ha
+  · simp only [dvd_zero]
+  rcases eq_or_ne b 0 with rfl | hb
+  · simp [factorization_lcm_left]
+  nth_rewrite 2 [← factorization_prod_pow_eq_self ha]
+  rw [prod_of_support_subset (s := (lcm a b).factorization.support)]
+  · apply prod_dvd_prod_of_dvd; rintro p -; dsimp only; split_ifs with le
+    · rw [factorization_lcm ha hb]; apply pow_dvd_pow; exact sup_le le_rfl le
+    · apply one_dvd
+  · intro p hp; rw [mem_support_iff] at hp ⊢
+    rw [factorization_lcm ha hb]; exact (lt_sup_iff.mpr <| .inl <| Nat.pos_of_ne_zero hp).ne'
+  · intros; rw [pow_zero]
+
+lemma factorization_lcm_right_dvd : factorization_lcm_right a b ∣ b := by
+  rcases eq_or_ne a 0 with rfl | ha
+  · simp [factorization_lcm_right]
+  rcases eq_or_ne b 0 with rfl | hb
+  · simp only [dvd_zero]
+  nth_rewrite 2 [← factorization_prod_pow_eq_self hb]
+  rw [prod_of_support_subset (s := (lcm a b).factorization.support)]
+  · apply Finset.prod_dvd_prod_of_dvd; rintro p -; dsimp only; split_ifs with le
+    · apply one_dvd
+    · rw [factorization_lcm ha hb]; apply pow_dvd_pow; exact sup_le (not_le.1 le).le le_rfl
+  · intro p hp; rw [mem_support_iff] at hp ⊢
+    rw [factorization_lcm ha hb]; exact (lt_sup_iff.mpr <| .inr <| Nat.pos_of_ne_zero hp).ne'
+  · intros; rw [pow_zero]
+
```

Comment:
Add
```suggestion

lemma div_gcd_mul_div_gcd_eq_lcm (a b : ℕ) :
    a / a.gcd (a / factorization_lcm_left a b) * (b / b.gcd (b / factorization_lcm_right a b)) =
      lcm a b := by
  rcases eq_or_ne a 0 with rfl | ha
  · simp only [Nat.zero_div, gcd_self, Nat.div_zero, zero_mul, lcm_zero_left]
  rcases eq_or_ne b 0 with rfl | hb
  · simp only [Nat.zero_div, gcd_self, Nat.div_zero, mul_zero, lcm_zero_right]
  have h₁ := factorization_lcm_left_pos a b
  have h₂ := factorization_lcm_right_pos a b
  obtain ⟨ka, hka⟩ := factorization_lcm_left_dvd a b
  obtain ⟨kb, hkb⟩ := factorization_lcm_right_dvd a b
  have hka' : ka ∣ a := ⟨factorization_lcm_left a b, mul_comm ka _ ▸ hka⟩
  have hkb' : kb ∣ b := ⟨factorization_lcm_right a b, mul_comm kb _ ▸ hkb⟩
  have hka₀ : 0 < ka := Nat.pos_of_ne_zero fun H ↦ ha <| Nat.zero_dvd.mp <| H ▸ hka'
  have hkb₀ : 0 < kb := Nat.pos_of_ne_zero fun H ↦ hb <| Nat.zero_dvd.mp <| H ▸ hkb'
  rw [Nat.div_eq_of_eq_mul_right h₁ hka, Nat.div_eq_of_eq_mul_right h₂ hkb,
    gcd_eq_right_iff_dvd.mp hka', gcd_eq_right_iff_dvd.mp hkb',
    Nat.div_eq_of_eq_mul_left hka₀ hka, Nat.div_eq_of_eq_mul_left hkb₀ hkb]
  exact factorization_lcm_left_mul_factorization_lcm_right ha hb
```
and use that below.

Reply:
(And similarly for the coprimality statement.)

Reply:
Maybe define abbreviations for `div_gcd_mul_div_gcd_eq_lcm` and `b / b.gcd (b / factorization_lcm_right a b)`?
Are these actually the same as `facotization_lcm_right b a` and `factorization_lcm_left b a`?
If so, it may be simpler to just use these later (and prove the relevant relations here).

Reply:
> Add
> 
> and use that below.

This is surely a general lemma unrelated to `factorization_lcm_left` and `factorization_lcm_right`, let me see.